agent/proxycfg/connect_proxy.go

```diff
@@ -155,11 +155,6 @@ func (s *handlerConnectProxy) initialize(ctx context.Context) (ConfigSnapshot, e
 		if u.Datacenter != "" {
 			dc = u.Datacenter
 		}
-		if s.proxyCfg.Mode == structs.ProxyModeTransparent && (dc == "" || dc == s.source.Datacenter) {
```

Comment:
Listeners in `xds/listeners.go` are spawned by the discovery chain watch that is initialized a few lines below this. I'm not entirely sure if introducing this discovery chain will cause other side-effects elsewhere in the application. It doesn't appear as if it will cause any issues.

Reply:
An issue with this approach is that this discovery chain for this service will eventually get deleted.

When the list of upstreams from intentions comes in we clean up any data for service names that weren't present in that response:
https://github.com/hashicorp/consul/blob/main/agent/proxycfg/connect_proxy.go#L466

However, virtual services from resolvers won't be returned by that watch. Intention upstreams are computed by starting from the list of service names in the catalog, and then filtering out based on whether intentions allow/deny access to that name. 

For this explicit upstream to persist you'll likely need to either:
* Update the reconciling logic in proxycfg to not delete data from explicit upstreams when we receive the list of intention upstreams
* Return service resolvers as valid intention upstreams

This issue doesn't show up in the new unit test because the `db-redir` discovery chain event happens after the `intentionUpstreamsID` update, but if the order was flipped, the output would change because of the reconciliation.